### PR TITLE
Folder And Dataset Metadata

### PR DESCRIPTION
**Pull request by MichaelBuessemeyer**
### URL of deployed dev instance (used for testing):
- https://folderdatasetproperties.webknossos.xyz/

### Steps to test:
- Open the dashboard & create another folder.
- Try adding metadata to the folder. Try out all three kinds of entries. Test whether autosave works
- Try editing metadata for a dataset
- Switch between datasets, the metadata should stay consistent
- Switch between folders, the metadata should stay consistent
- And also switch between folders and datasets and also look for any errors / inconsistencies / lost metadata entries.

### TODOs:
- [x] Clean up git history 
- [x] Allow to search for metadata props -> Other issue https://github.com/scalableminds/webknossos/issues/8034
- [x] Add props & UI and everything for folders 
- [x] Update UI to the newest design proposal
- [x] Clean up code
- [x] Test dataset update route without supplying the metadata field 
  - [x] Works. Tested with firefox's edit and resend functionality. 

### Open Question
- [x] What about supporting dataset tags? Should they be removed? Should one still be able to update the dataset tags via the API? Should they be removed from the schema? Should they be transferred into metadata entries?
- [x] How should tags be transferred into metadata?
  - As a single string array entry with key "tags"
  - For each on entry with numbered keys like "tagX" and x is the index of the tag in the tag array
  - other ideas 

### Issues:
- fixes [#5968](https://github.com/scalableminds/webknossos/issues/5968)

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [x] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [x] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [x] Needs datastore update after deployment

### CODE REVIEW

**Comment by MichaelBuessemeyer** [comment]
For the record:

During development / testing, I noticed an error with the antd Select component used in the new MetadataTable. Looks like this is this one of the related issues: https://github.com/ant-design/ant-design/issues/46348. I just gave it a quick look. The error might be gone after a proper antd & react update :)

image

**Comment by hotzenklotz** [comment]
Things I noticed during testing:
- [ ] The type `a,b` always has an empty tag as a default
  Screenshot 2024-07-17 at 11 23 30

- [ ] No matter which dataset I selected from my dashboard they all had the same values 
  Screenshot 2024-07-17 at 11 27 29


Some UI/UX feedback:
- [ ] The colon between "property" and "value" is not centered. Better yet, we can remove it. 
- [ ] The properties are always displayed as text boxes even when one does not want to edit them. Text boxes are not great for readability. Maybe make the inputs `borderless` when they are not actively being edited (https://ant.design/components/input#components-input-demo-variant)
- [ ] Similar to the comment above, there is very little space in general. Maybe the "type" dropdown" and the "delete" button don't always have to be there. They are only relevant when working/editing an entry. Similar to other UI elements in WK every entry can enter an edit mode where all of that stuff is displayed. When not in editing mode, then the properties are just displayed as text. To avoid a pencil icon, one could trigger the edit mode when clicking on the text similiar to the antd `<Typography.Text>` component (https://ant.design/components/typography#components-typography-demo-editable). 
- [ ] Have a look a the antd `<Space.Compact>` component for clean compact input fields (https://ant.design/components/space#space-demo-compact). e.g. 
  Screenshot 2024-07-17 at 11 57 18
- [ ] The "types" field is only relevant for the values. All keys are always strings anyway. Therefore the "type" dropdown should can be right next to the "value" input and need not be the first entry in the table. (see comment directly above)
- [ ] I am not a big fan of the "plus" button to add new entry. Especially the empty state looks weird with just the table header and button. Maybe always show and empty to row, ready to edit?
- [ ] I wonder if the "types" options could be better explained through icons. Especially the multiple values/tags thing is not obvious. Antd has several options that might make sense, e.g. `<TagsOutlined />` 
  Screenshot 2024-07-17 at 11 38 27 Screenshot 2024-07-17 at 11 40 52

**Comment by MichaelBuessemeyer** [comment]
>  The type a,b always has an empty tag as a default

Thanks, I remove fix that

> No matter which dataset I selected from my dashboard they all had the same values

Afaik, these are default added values for a new dataset. I can change that in the backend. You can change them individually for each dataset and the should be shown and saved individually.

>The colon between "property" and "value" is not centered. Better yet, we can remove it.

The colon was explicitly requested to be added. I centered it now.

> The properties are always displayed as text boxes even when one does not want to edit them. Text boxes are not great for readability. Maybe make the inputs borderless when they are not actively being edited (https://ant.design/components/input#components-input-demo-variant)

The inputs were already borderless. But I now also made the background transparent when the entry is not being edited.

>  Similar to the comment above, there is very little space in general. Maybe the "type" dropdown" and the "delete" button don't always have to be there. They are only relevant when working/editing an entry. Similar to other UI elements in WK every entry can enter an edit mode where all of that stuff is displayed. When not in editing mode, then the properties are just displayed as text. To avoid a pencil icon, one could trigger the edit mode when clicking on the text similiar to the antd <Typography.Text> component (https://ant.design/components/typography#components-typography-demo-editable).
> Have a look a the antd <Space.Compact> component for clean compact input fields (https://ant.design/components/space#space-demo-compact). e.g.

Thanks something to be discussed (I'd say as I got a different design proposal).

> The "types" field is only relevant for the values. All keys are always strings anyway. Therefore the "type" dropdown should can be right next to the "value" input and need not be the first entry in the table. (see comment directly above)

The whole type dropdown was removed to save some space in the table. Users can now select a type when creating a new entry and that type will no longer be changable after creation.

> I am not a big fan of the "plus" button to add new entry. Especially the empty state looks weird with just the table header and button. Maybe always show and empty to row, ready to edit?


Initially there always was an empty row. But it was decided against this. See https://scm.slack.com/archives/C5AKLAV0B/p1719923640762759 and https://scm.slack.com/archives/C5AKLAV0B/p1718896223081789

> I wonder if the "types" options could be better explained through icons. Especially the multiple values/tags thing is not obvious. Antd has several options that might make sense, e.g. <TagsOutlined />

Thanks, I did know about these icons. I added them to the newest version.

File: app/models/dataset/DatasetService.scala
Comment (fm3) [inline app/models/dataset/DatasetService.scala:83]: Why is the "index" needed? If the data model is an array, the ordering should already be fixed, right?

File: CHANGELOG.unreleased.md
Comment (fm3) [inline CHANGELOG.unreleased.md:15]: Maybe rename the PR to also use the word “Metadata” instead of Properties? (Or update the code to also use Properties) I’d like to introduce only one new term, not two :sweat_smile:

File: app/models/folder/Folder.scala
Comment (fm3) [inline app/models/folder/Folder.scala:141]: this might already be worth a one-liner helper function `private def parseMetadata(literal: Option[String]): JsArray)` :)

**Review by fm3 — COMMENTED** [review]
Backend looking good already. Let’s decide on a terminology (metadata vs properties).

Also, since the dataset json changes in the API (details field is replaced by metadata field with different structure), we need to take special care. This PR adapts the frontend. The python libs don’t use the details field as far as I can see. @normanrz My guess is that the publications rendering code from the about page of webknossos needs to be adapted. Do you think that would be straightforward? Also do you know if external collaborators rely on the details field in the dataset json? If so, we would need to bump the API version.

**Comment by normanrz** [comment]
> My guess is that the publications rendering code from the about page of webknossos needs to be adapted. Do you think that would be straightforward?

Yes, I can take of this once this PR is deployed.

File: app/models/dataset/DatasetService.scala
Comment (MichaelBuessemeyer) [reply app/models/dataset/DatasetService.scala:83]: > I could remove the index and always re-add it in the frontend 🤔

Its now implemented that way

**Comment by MichaelBuessemeyer** [comment]
@philippotto I prepared the code ready for your feedback and testing :D.

Moreover, I just removed an unnecessary dependency of a `useEffect` hook introduced by me in another pr which was just merged today.

**Comment by MichaelBuessemeyer** [comment]
@markbader Metadata should now be updateable via the normal dataset update route

**Review by philippotto — COMMENTED** [review]
haven't looked at the code yet, but testing results are in :)

image

- [x] the table head ("Property") and the validation warning are not aligned with the rest of the table entries. strictly speaking, the input box is also a bit too wide, but since the text within that box is aligned, it feels good.
- [ ] the trash icon is incorrectly positioned when a validation warning is there
- [ ] auto-save also happens when a field is focussed. after auto-saving that focus is not there anymore. it's probably easiest to not auto-save as long as a field is focussed.

File: frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx
Comment (philippotto) [reply frontend/javascripts/dashboard/advanced_dataset/dataset_table.tsx:369]: hmm, good catch. it's definitely a bit fishy.

the logic should be as follows:
- if the table is sorted by a certain column (columnKey is properly defined), let antd sort the table accordingly (I think, this happens automatically)
- if not:
  - if no search is active: sort by lastUsedByUser
  - if a search is active: sort by dice coefficient

can you doublecheck that it works now correctly?

> although this should be done by the backend nowadays, or am I wrong?

no, I think, the frontend should do the sorting.

File: frontend/javascripts/dashboard/dataset/queries.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/dataset/queries.tsx:327]: I wonder whether these two lines should be above `...updatedFolder` so that they don't have precedence. however, moving a folder and updating the metadata seem to work currently :thinking:

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:64]: normal if would be more idiomatic

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:91]: same here

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:99]: same here

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:110]: I'd move the types to the top of the module

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:128]: can be moved into effect. I'd probably even inline it.

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:153]: maybe extract into helper since it is needed multiple times?

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:132]: not sure whether I understand it correctly: this metadata table is not remounted when activating another dataset/folder item in the list, right? therefore, you check whether the item changed in this effect. if yes, you flush the pending updates. is this correct? I think, it would be cleaner to unmount and remount the component when the dataset/folder changed. that way, the component's state is perfectly isolated (right now; I'd be a bit afraid that the metadata slips into the other dataset/folder in edge cases). the "use previous" code here wouldn't be necessary, either, I think.
you could achieve this easily by passing a `key` like this `<MetadataTable key={"${item.name}-${item.type}"} />`.
or was there a reason for not doing this?

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:56]: I'd rename dontUpdateState to `isComponentMounted` and add the comment that this is necessary to avoid the set-functions if it's not.
another idea: you could even wrap the set functions in the other component and adapt them so that they don't do anything if its not mounted. I think that would be cool because the wrapping would only happen once. currently, you need to define and update this signal once per useEffect definition.

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:182]: why isn't this equivalent to `prev[index]`?

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:190]: since both set the error I'd turn this into an if-else.

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:202]: is there a reason for moving the item to the back? intuitively, I would have mapped the items and only update the one with the requested index (even though I'm still not sure I get the relevance of the index; isn't it redundant?).

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:215]: same here

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:295]: the switch block could be a good candidate for an own functional component.

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:409]: own component? :)

**Review by philippotto — CHANGES_REQUESTED** [review]
I reviewed the code now. Cool stuff :) I left a couple of nitpicks. on a higher level I have two things (I commented both with more detail).

1. I don't really get the index property of the metadata items. wouldn't it work without it?
2. I'd enforce a remount when the user clicks on another folder/dataset, unless there is a strong reason against it.

it's probably best to discuss this in person before writing long walls of text :) :call_me_hand:

**Review by MichaelBuessemeyer — COMMENTED** [review]
Thanks for all the feedback @philippotto 

Everything should be incorporated now. Please have another look :)

File: frontend/javascripts/dashboard/folders/details_sidebar.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/details_sidebar.tsx:290]: please add a comment that the key is crucial to enforce remounts when the id changes.

File: frontend/javascripts/dashboard/folders/details_sidebar.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/details_sidebar.tsx:183]: see below

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:145]: the comment can now be removed thanks to the wrapped setIsSaving, I think

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (philippotto) [inline frontend/javascripts/dashboard/folders/metadata_table.tsx:251]:
```suggestion
  // Send automatic async debounced updates to the server when metadata changed and there are no focused rows.
```

File: frontend/javascripts/dashboard/folders/details_sidebar.tsx
Comment (MichaelBuessemeyer) [reply frontend/javascripts/dashboard/folders/details_sidebar.tsx:290]: I also added a comment to the component itself

File: frontend/javascripts/dashboard/folders/metadata_table.tsx
Comment (MichaelBuessemeyer) [reply frontend/javascripts/dashboard/folders/metadata_table.tsx:145]: yeah right, thanks for paying attention :pray:

**Review by MichaelBuessemeyer — COMMENTED** [review]
Thanks for the fast review :pray: . I implemented your feedback :+1:

**Review by philippotto — COMMENTED** [review]
front-end looks good :+1: one visual nitpick:

image

the trashcans should have a margin-right or something like that. there's very little room right now.

also, the PR has two open questions which need to be addressed. I don't have a strong opinion about them. maybe ask for opinions in slack?

**Comment by MichaelBuessemeyer** [comment]
Before merging the migration & revision needs to be tested. I already tested this for me locally by:
- check out master
- run `yarn refresh-schema` 
- open wk dashboard
- Add some tags for at least one dataset and remember them
- Open the datasets table of the Postgres db manually and add some "species", "acquisition", and "brainRegion" to the details like done for publications
- Then checkout this branch
- Run the migration on the db -> this should succeed
- open the wk dataset dashboard and take a look at the metadata of the datasets that previously had tags. They should now be transferred to a tags entry
- use the dbtool to run a schema diff. This should run successfully without showing any schema diffs
- run the revision SQL script
- checkout master
- run schema diff -> should also show no schema diff
- Check the dataset dashboard: The previously added tags should appear there

**Comment by philippotto** [comment]
the front-end looks good to me :+1: leaving final approval to @fm3. he probably should have a look at the migration code, too. also, does the python lib need to be adapted to this pr?

**Comment by MichaelBuessemeyer** [comment]
>he probably should have a look at the migration code, too. 

Yeah definitely.

Moreover: as the tags should be deprecated should they be remove from the schema and API objects returned by the API or should they stay but the frontend cannot update and view them?

**Review by fm3 — CHANGES_REQUESTED** [review]
Backend looking good already :) I added a couple of comments.
I also tested a bit more and it’s looking pretty good :)

I did not yet review the migrations. I think we need to decide about the NULLability first, and also about what to do with the tags.

If I understand correctly, the dataset tags are now turned into the new metadata. If this is true, the tags should be removed from the SQL schema and the Dataset case class.

However, the dataset json still needs to serve the tags for the moment, until the libs client is adapted. I think it would be good if the old tags could be served there. Meaning that the json serialization should convert the relevant metadata to tags and serve them, alongside the new metadata attribute. In the same way, updating the tags from the libs should also stay possible, so an adapter for that is needed as well.

Also, could you create a libs issue to adapt the libs to the new metadata format and deprecate the tags there too?

Also note that https://github.com/scalableminds/webknossos/pull/8006 was merged first so that the schema version in this PR needs to be increased to avoid conflicts.

File: tools/postgres/schema.sql
Comment (MichaelBuessemeyer) [reply tools/postgres/schema.sql:126]: :thinking: My thought here was that it might be easier to work with at least an empty array instead of null. I added the null / None support to the code to ensure easier compatibility. This includes update calls not necessarily having to pass metadata (at least I hope that's the case, as I wanted to implement it that way).

But I can easily remove the default value from the column :) Or even go further and make metadata non-optional if you think this is preferable.

**Comment by MichaelBuessemeyer** [comment]
Ok what to do with tags :)

I just talked with @normanrz and the conclusion is to **keep tags** as some are using them as well as the pretty important filtering feature.

If we later want to merge the tags into the metadata, we can still do that. In case dataset/folder already has a metadata field named "tags" already exists, we just come up with something :).

File: frontend/javascripts/dashboard/folders/details_sidebar.tsx
Comment (MichaelBuessemeyer) [inline frontend/javascripts/dashboard/folders/details_sidebar.tsx:180]: Open question? Is it ok to move the tags into the "spinner area"? Previously they weren'T but, the metadata should be within the "spinner area" as it depends on `fullDataset` and IMO the newly added UI element should appear after the already known tags.

Comment (philippotto) [reply frontend/javascripts/dashboard/folders/details_sidebar.tsx:189]: sure, why not :+1:

**Comment by MichaelBuessemeyer** [comment]
testing went well. Tags should be restored in the frontend :)

File: tools/postgres/schema.sql
Comment (fm3) [reply tools/postgres/schema.sql:126]: I don’t really have a preference for default value versus nullable. But I don’t much like having both. Please pick what’s easier implementation-wise :) My guess is that removing the default value (and so have the default be NULL/None) is easier.

Comment (MichaelBuessemeyer) [reply tools/postgres/schema.sql:126]: Not is should be non nullable and with `DEFAULT '[]` for dataset and folders :).

The update route for datasets parses the metadata field as optional and used the already existing metadata in the backend query in case the user / client did not provide any metadata (untested) :see_no_evil:

**Review by fm3 — COMMENTED** [review]
Backend LGTM

**Review by philippotto — APPROVED** [review]
just tested the dev instance again and everything looks good :+1: I didn't test the migration. ifi that is tested well, feel free to merge and deploy :shipit:

File: conf/evolutions/reversions/119-add-metadata-to-folders-and-datasets.sql
Comment (MichaelBuessemeyer) [inline conf/evolutions/reversions/119-add-metadata-to-folders-and-datasets.sql:17]: The newest commit changed 
```
SET details = jsonb_set(details, '{species}', (
```
to
```
SET details = jsonb_set(COALESCE(details, '{}'), '{species}', (
```

This was necessary, as the default value of the newly re-added details column was null and thus the `jsonb_set` would also return null. This should work now :)

**Review by MichaelBuessemeyer — COMMENTED** [review]
@fm3 I just pushed a commit to fix the reversion. Could you please give me a final thumbs up?

I just tested the migration and reversion. both should work

**Comment by MichaelBuessemeyer** [comment]
And @fm3 no need to test the migration again. I tested it and it should work :tm: :see_no_evil:

**Pull request merged**
